chainstate/src/detail/spend_cache.rs

```diff
@@ -259,27 +306,56 @@ impl<'a> CachedInputs<'a> {
         passes; the check is down there. Feel free to move stuff around.
         */
 
-        let tx = block
-            .transactions()
-            .get(tx_num)
-            .ok_or_else(|| BlockError::TxNumWrongInBlock(tx_num, block.get_id()))?;
+        match spend_kind {
+            SpendKind::Transaction(tx_num) => {
+                let tx = block
+                    .transactions()
+                    .get(tx_num)
+                    .ok_or_else(|| BlockError::TxNumWrongInBlock(tx_num, block.get_id()))?;
 
-        // pre-cache all inputs
-        tx.get_inputs()
-            .iter()
-            .try_for_each(|input| self.fetch_and_cache(input.get_outpoint()))?;
+                // pre-cache all inputs
+                tx.get_inputs()
+                    .iter()
+                    .try_for_each(|input| self.fetch_and_cache(input.get_outpoint()))?;
+
+                // check for attempted money printing
+                self.check_inputs_amounts(tx.get_inputs(), tx.get_outputs())?;
 
-        // check for attempted money printing
-        self.check_inputs_amounts(tx)?;
+                // verify input signatures
+                self.verify_signatures(tx)?;
 
-        // verify input signatures
-        self.verify_signatures(tx)?;
+                // spend inputs of this transaction
+                let spender = Spender::from(tx.get_id());
+                self.apply_spend(tx.get_inputs(), spend_height, blockreward_maturity, spender)?;
+            }
+            SpendKind::BlockReward => {
+                match block.header().block_production_inputs() {
+                    Some(inputs) => {
+                        // pre-cache all inputs
+                        inputs
+                            .iter()
+                            .try_for_each(|input| self.fetch_and_cache(input.get_outpoint()))?;
+
+                        // check for attempted money printing
+                        if let Some(outputs) = block.header().block_reward_destinations() {
+                            self.check_inputs_amounts(inputs, outputs)?;
+                        }
+                    }
+                    None => {}
+                }
 
-        // spend inputs of this transaction
-        self.apply_spend(tx, spend_height, blockreward_maturity)?;
+                // verify input signatures
+                self.verify_all_signatures(block)?;
```

Comment:
I'm having trouble understanding this. So this block reward attempts to spend some output. And to check whether the signature is correct, hence, the spend is authorized, we check all the signatures in the block except for the one we're checking?! This is definitely incorrect.